### PR TITLE
Speed improvements

### PR DESCRIPTION
**Pull request by eregon**
This PR improves the speed of the emulator on Tetris from 8 FPS to 30 FPS (still some way to go for 60).

I used [stackprof](https://github.com/tmm1/stackprof) to identity the hotspots.
The sample profiler is now enabled by passing `--stackprof` to `exe/waterfoul start`.

The emulator now also displays the FPS in the window title.

The initial stackprof report was:
```
==================================
  Mode: cpu(1000)
  Samples: 119330 (0.01% miss rate)
  GC: 3109 (2.61%)
==================================
     TOTAL    (pct)     SAMPLES    (pct)     FRAME
     90270  (75.6%)       84166  (70.5%)     Waterfoul::MMU#[]
    129160 (108.2%)        7744   (6.5%)     Waterfoul::PPU#render_bg
      3912   (3.3%)        3912   (3.3%)     Waterfoul::Cartridge#[]
      2911   (2.4%)        2911   (2.4%)     Waterfoul::PPU#rgb
      1736   (1.5%)        1736   (1.5%)     Waterfoul::MBC::ROM#[]
      1850   (1.6%)        1584   (1.3%)     Waterfoul::MMU#[]=
     14277  (12.0%)        1474   (1.2%)     Waterfoul::Interrupt.pending_interrupt
      1177   (1.0%)        1177   (1.0%)     Waterfoul::CPU#reset_tick
```
So `MMU#[]` was clearly the bottleneck.

After these commits it looks like:
```
==================================
  Mode: cpu(1000)
  Samples: 35996 (0.01% miss rate)
  GC: 386 (1.07%)
==================================
     TOTAL    (pct)     SAMPLES    (pct)     FRAME
      5177  (14.4%)        5177  (14.4%)     Waterfoul::MMU#read_memory_byte
      7575  (21.0%)        4854  (13.5%)     Waterfoul::MMU#[]
      9465  (26.3%)        4307  (12.0%)     Waterfoul::PPU#render_bg
      3172   (8.8%)        3172   (8.8%)     Waterfoul::PPU#rgb
      2627   (7.3%)        2627   (7.3%)     Waterfoul::MBC::ROM#[]
      3166   (8.8%)        1770   (4.9%)     Waterfoul::Interrupt.pending_interrupt
      1830   (5.1%)        1598   (4.4%)     Waterfoul::MMU#[]=
      1121   (3.1%)        1121   (3.1%)     Waterfoul::CPU#reset_tick
     15416  (42.8%)        1112   (3.1%)     Waterfoul::PPU#step
       838   (2.3%)         838   (2.3%)     Waterfoul::CPU#instruction_cycle_time
      2379   (6.6%)         823   (2.3%)     Waterfoul::Timer#tick
      8400  (23.3%)         821   (2.3%)     Waterfoul::CPU#perform_instruction
       743   (2.1%)         743   (2.1%)     Waterfoul::Instructions::Registers#set_z_flag
       732   (2.0%)         732   (2.0%)     Waterfoul::Instructions::Registers#reset_flags
```
It is less clear, but `PPU#render_bg` seems fairly slow compared to the rest.
Maybe some higher-level optimization could be done such as pre-computing a recurring pattern once, or if the background is a uniform color that might be easy to detect.

### CODE REVIEW

File: lib/waterfoul/mmu.rb [hunk 1]
```diff
@@ -33,26 +30,14 @@ def initialize
...
         Input.read_keyboard @memory[i]
-      when 0x0000...0x8000 # ROM Bank 0 + n
-        # if the boot rom is enabled and the address is < 0x100
-        if @map_boot_rom && i <= BOOT_ROM_END_MEM_LOC
-          BootROM[i]
-        else
-          @cartridge[i]
-        end
-      when 0x8000...0xA000 # Video RAM
+      elsif i >= 0x8000 && i < 0xA000 or i >= 0xC000 && i < 0xE000 or i >= 0xFE00
```
Comment (eregon) [inline lib/waterfoul/mmu.rb]: This is less nice than the ranges, but avoiding their allocation is a major speedup on the standard ruby/MRI.

**Comment by colby-swandale** [comment]
Holy wow this is amazing, i can actually play a reasonable game of Tetris. This also really speeds up Mario and Pokemon as well.

I'm going to be looking at this over the next few days and probably adding my stuff as well.

Thanks again!!! 🎉

File: lib/waterfoul/ppu.rb [hunk 2]
```diff
@@ -206,14 +206,14 @@ def render_sprites
       39.downto(0) do |sprite|
         sprite_offset = sprite * 4
 
-        sprite_y = $mmu.read_byte(0xFE00 + sprite_offset) - 16
+        sprite_y = $mmu.read_memory_byte(0xFE00 + sprite_offset) - 16
```
Comment (eregon) [inline lib/waterfoul/ppu.rb]: These `read_memory_byte` seem to improve by a few FPS (< 5 out of 30), so it is some kind of trade-off.

**Comment by colby-swandale** [comment]
sorry for the delay, i'll probably spend this weekend going through this PR. 😄

File: lib/waterfoul/boot_rom.rb [hunk 3]
```diff
@@ -49,10 +49,19 @@ class BootROM
       0x3E, 0x01, 0xE0, 0x50
     ].freeze
 
+    attr_reader :cartridge
```
Comment (colby-swandale) [inline lib/waterfoul/boot_rom.rb]: I'm not comfortable about moving this into the `BootRom`. Is there anyway we can move can we keep these separate while keeping the performance gains?

File: lib/waterfoul/ppu.rb [hunk 4]
```diff
@@ -37,19 +37,19 @@ def step(cycles = 1)
...
-        when V_BLANK_STATE
-          vblank
-        when OAM_READ_STATE
-          oam if @modeclock >= OAM_SCANLINE_TIME
-        when VMRAM_READ_STATE
-          vram if @modeclock >= VRAM_SCANLINE_TIME
-        end
-      else
-        if IO::LCDControl.screen_enabled?
+      if IO::LCDControl.screen_enabled?
```
Comment (colby-swandale) [inline lib/waterfoul/ppu.rb]: What's the reason for this change?

File: lib/waterfoul/boot_rom.rb (hunk 3 above)
Comment (eregon) [inline lib/waterfoul/boot_rom.rb]: I thought it's a nice way to model how the memory lookup works while booting: low addresses are overridden by the boot rom, but higher addresses access the real rom. So this "chains" them nicer than two `if` in MMU.
Also, this allows just replacing the rom by the real cartridge once boot is finished, instead of maintaining an extra `@map_boot_rom` state in MMU.

Comment (eregon) [inline lib/waterfoul/boot_rom.rb]: There could be a third abstraction on top of the boot rom and the cartridge, but I am not sure it is better.
Essentially, the boot rom alone is not enough for booting, a real cartridge is needed as well.

File: lib/waterfoul/ppu.rb (hunk 4 above)
Comment (eregon) [inline lib/waterfoul/ppu.rb]: It avoids calling `IO::LCDControl.screen_enabled?` and slightly simplifies the control flow.

Comment (eregon) [inline lib/waterfoul/ppu.rb]: Compare https://github.com/eregon/waterfoul/blob/f095a087a8e9d1f64eb105b2e4caff4b2d4e4b8b/lib/waterfoul/ppu.rb#L40-L64 and https://github.com/eregon/waterfoul/blob/18e62d17221644a79aa88b3b7f9952ea8803a75d/lib/waterfoul/ppu.rb#L40-L64

**Comment by colby-swandale** [comment]
Sorry for the delay again, i'm pretty happy with this merge. Thanks again!

**Comment by eregon** [comment]
You're welcome, thanks for merging!
